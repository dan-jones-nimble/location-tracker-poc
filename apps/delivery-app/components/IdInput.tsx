// import { useRouter } from 'expo-router';
import { StyleSheet, TextInput, View, Text } from 'react-native';
import { getLocationPings } from '@nx-expo/appwrite';
import { useState, useEffect } from 'react';
import { LocationObject } from 'expo-location';
import getDirections from 'react-native-google-maps-directions';

export const IdInput = () => {
  // const router = useRouter();
  const [locationResponse, setLocationResponse] =
    useState<Array<LocationObject>>();

  const handleTextChange = (routeId: string) => {
    if (routeId.length >= 36) {
      getLocationPings(routeId).then((pings) => {
        // router.push(`/jobList?routeId=${routeId}`);
        setLocationResponse(pings);
      });
    }
  };

  useEffect(() => {
    if (locationResponse && locationResponse?.length > 0) {
      console.log("location response: " + JSON.stringify(locationResponse));
      const waypoints = locationResponse;
      const firstWaypoint = waypoints.shift()!;
      const lastWaypoint = waypoints.pop()!;
      const remainingWaypoints = waypoints.map(waypoint => {
        console.log("waypoint: " + JSON.stringify(waypoint));
        return {
          latitude: waypoint.latitude,
          longitude: waypoint.longitude
        }
      });
      const data = {
        source: {
          latitude: firstWaypoint.latitude,
          longitude: firstWaypoint.longitude
        },
        destination: {
          latitude: lastWaypoint.latitude,
          longitude: lastWaypoint.longitude
        },
        params: [
          {
            key: "travelmode",
            value: "driving"
          },
          {
            key: "dir_action",
            value: "navigate"
          }
        ],
        waypoints: remainingWaypoints
      }
      getDirections(data);
    }
  }, [locationResponse]);

  return (
    <View style={styles.container}>
      {locationResponse ? (
        <Text>{JSON.stringify(locationResponse)}</Text>
      ) : (
        <TextInput
          style={styles.textInput}
          inputMode="numeric"
          placeholder="123456"
          autoFocus
          maxLength={36}
          onChangeText={(text) => handleTextChange(text)}
        />
      )}
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  textInput: {
    display: 'flex',
    fontSize: 16
  }
});
