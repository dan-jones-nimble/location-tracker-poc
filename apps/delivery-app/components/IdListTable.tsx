import { MinimalLocationObject, RouteObject } from "@nx-expo/appwrite";
import { openURL } from "expo-linking";
import { View, Text, StyleSheet, Pressable } from "react-native";

const buildWaypointsUrl = (waypoints: Array<MinimalLocationObject>) => {
	const firstWaypoint = waypoints.shift();
	const lastWaypoint = waypoints.pop();
	const remainingWaypoints = waypoints.map(
		(locationObject: MinimalLocationObject) => ({
			latitude: locationObject.latitude,
			longitude: locationObject.longitude
		})
	);

	const originParam = `origin=${firstWaypoint.latitude},${firstWaypoint.longitude}`;
	const destinationParam = `destination=${lastWaypoint.latitude},${lastWaypoint.longitude}`;
	const waypointsParam = `waypoints=${remainingWaypoints
		.map((waypoint) => `${waypoint.latitude},${waypoint.longitude}`)
		.join('|')}`;

	return `https://www.google.com/maps/dir/?api=1&${destinationParam}&${originParam}&${waypointsParam}`;
};

export const IdListTable = (props: {
	routes: Array<RouteObject>
}) => {
	const handleButtonPress = (route: RouteObject) => {
		const gMapsUrl = buildWaypointsUrl(route.location_pings);
		openURL(gMapsUrl).catch((error) => console.error(error));
	};

	return (
		<View style={styles.tableView}>
			{props.routes.map((route) => (
				<Pressable onPress={() => handleButtonPress(route)}>
					<Text style={styles.text} key={route.$id}>{route.$id}</Text>
				</Pressable>
			))}
		</View>
	);
}

const styles = StyleSheet.create({
	tableView: {
		display: "flex",
		gap: 20,
		alignItems: "center",
		justifyContent: "center",
	},
	text: {
		fontSize: 16,
		lineHeight: 16,
	},
});