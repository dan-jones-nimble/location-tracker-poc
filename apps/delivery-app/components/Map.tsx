import { RouteObject } from '@nx-expo/appwrite';
import MapView, { Marker, Polyline, Provider, Region } from 'react-native-maps';
import { StyleSheet, useWindowDimensions } from 'react-native';
import { useMemo, useState } from 'react';

interface MapProps {
  route: RouteObject;
  provider: Provider;
}

export const Map = ({ route, provider }: MapProps) => {
  const locationPings = route.location_pings;
  const { height, width } = useWindowDimensions();
  const ASPECT_RATIO = useMemo(() => width / height, [width, height]);
  const [currentRegion, setCurrentRegion] = useState<Region>({
    latitude: locationPings[0].latitude,
    longitude: locationPings[0].longitude,
    latitudeDelta: 0.01,
    longitudeDelta: 0.01 * ASPECT_RATIO
  });

  return (
    <MapView
      style={StyleSheet.absoluteFill}
      initialRegion={currentRegion}
      onRegionChange={(region) => setCurrentRegion(region)}
      provider={provider}
      loadingEnabled
      showsUserLocation
      showsPointsOfInterest={false}
    >
      <Polyline
        coordinates={locationPings.map((locationPing) => ({
          latitude: locationPing.latitude,
          longitude: locationPing.longitude
        }))}
      />
      {locationPings.map((locationPing) => (
        <Marker
          key={locationPing.$id}
          coordinate={{
            longitude: locationPing.longitude,
            latitude: locationPing.latitude
          }}
        />
      ))}
    </MapView>
  );
};
