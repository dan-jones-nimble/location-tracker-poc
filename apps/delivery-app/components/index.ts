export { IdList } from './IdList';
export { IdListTable } from './IdListTable';
export { Map } from './Map';
export { MapModal } from './MapModal';
