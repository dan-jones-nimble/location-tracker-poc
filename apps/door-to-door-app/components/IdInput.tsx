/* eslint-disable jsx-a11y/accessible-emoji */
import { useRouter } from 'expo-router';
import { StyleSheet, TextInput, View } from 'react-native';

export const IdInput = () => {
  const router = useRouter();

  const submitRouteId = (jobNumber: string) =>
    new Promise((res) => res(() => console.log(jobNumber)));

  const handleTextChange = (routeId: string) => {
    if (routeId.length >= 6) {
      submitRouteId(routeId).then(() =>
        router.push(`/jobList?routeId=${routeId}`)
      );
    }
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.textInput}
        inputMode="numeric"
        placeholder="123456"
        autoFocus
        maxLength={6}
        onChangeText={(text) => handleTextChange(text)}
      />
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textInput: {
    display: 'flex',
    fontSize: 16,
  },
});
