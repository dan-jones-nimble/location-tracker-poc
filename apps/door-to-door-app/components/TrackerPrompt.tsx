// Imports
import { useEffect, useState } from 'react';
import { Text } from 'react-native';
import {
  requestPermissions,
  startLocationTracking,
  stopLocationTracking
} from '@nx-expo/location';

export const TrackerPrompt = () => {
  const [locationTrackingPermitted, setLocationTrackingPermitted] =
    useState<boolean>(false);
  const [text, setText] = useState<string>('You must accept tracking');

  const handleError = (error) =>
    setText(`You aren't being tracked...\n${error.message}`);

  useEffect(() => {
    requestPermissions()
      .then(() => setLocationTrackingPermitted(true))
      .catch(handleError);
  }, []);

  useEffect(() => {
    if (locationTrackingPermitted) {
      startLocationTracking()
        .then(() => setText('You are being tracked...'))
        .catch(handleError);
    }

    return () => {
      stopLocationTracking().then(() => setText("You aren't being tracked..."));
    };
  }, [locationTrackingPermitted]);

  return <Text>{text}</Text>;
};
