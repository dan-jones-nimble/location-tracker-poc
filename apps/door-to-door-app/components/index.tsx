export * from './IdInput';
export * from './TrackerPrompt';
