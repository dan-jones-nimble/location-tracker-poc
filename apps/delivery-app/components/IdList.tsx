import { IdListTable } from './IdListTable';
import { Text } from 'react-native';
import { RouteObject } from '@nx-expo/appwrite';

interface IdListProps {
  routes: Array<RouteObject>;
  onIdClick: (routeId: string) => void;
}

export const IdList = ({ routes, onIdClick }: IdListProps) => {
  return (
    <>
      {routes.length > 0 && (
        <IdListTable onIdClick={onIdClick} routes={routes} />
      )}
      {routes.length === 0 && <Text>No routes found</Text>}
    </>
  );
};
