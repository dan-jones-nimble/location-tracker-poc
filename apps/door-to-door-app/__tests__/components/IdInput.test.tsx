import { render, fireEvent, waitFor, screen } from '@testing-library/react-native';
import { IdInput } from '../../components/IdInput';

const mockPush = jest.fn();
jest.mock("expo-router", () => ({
	useRouter: () => ({
		push: mockPush,
	}),
}));

describe("IdInput", () => {
	it("matches snapshot", () => {
		const tree = render(<IdInput />);
		expect(tree).toMatchSnapshot();
	});

	it("text change, route id length is less than 6, router push is not called", () => {
		const text = "12345";
		const tree = render(<IdInput />);
		const textInput = tree.getByPlaceholderText("123456");
		fireEvent.changeText(textInput, { target: { value: text } });
		expect(mockPush).not.toHaveBeenCalled();
	});

	it("text change, route id length is 6, router push is called", async () => {
		const text = "123456";
		const tree = render(<IdInput />);
		const textInput = tree.getByPlaceholderText("123456");
		fireEvent.changeText(textInput, text);
		await waitFor(() => expect(mockPush).toBeCalledWith(`/jobList?routeId=${text}`));
		expect(tree).toMatchSnapshot();
	});

	it("text change, route id length is greater than 6, router push is called", async () => {
		const text = "abcdefghijklmno";
		const tree = render(<IdInput />);
		const textInput = tree.getByPlaceholderText("123456");
		fireEvent.changeText(textInput, { target: { value: text } });
		await waitFor(() => expect(mockPush).toBeCalledWith(`/jobList?routeId=${text}`));
		expect(tree).toMatchSnapshot();
	});

	it("text change, route id is empty, router push is not called", () => {
		const text = "";
		const tree = render(<IdInput />);
		const textInput = tree.getByPlaceholderText("123456");
		fireEvent.changeText(textInput, { target: { value: text } });
		expect(mockPush).not.toHaveBeenCalled();
	});

});