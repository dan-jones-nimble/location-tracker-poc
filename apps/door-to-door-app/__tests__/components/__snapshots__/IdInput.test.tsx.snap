// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdInput matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    autoFocus={true}
    inputMode="numeric"
    maxLength={6}
    onChangeText={[Function]}
    placeholder="123456"
    style={
      {
        "display": "flex",
        "fontSize": 16,
      }
    }
  />
</View>
`;
